---
title: "Supplementary Information: *Rtpca: an R package for differential thermal coaggregation analysis*"
author:
- name: Nils Kurzawa
  affiliation: 
  - European Molecular Biology Laboratory (EMBL), Genome Biology Unit
- name: AndrÃ© Mateus 
  affiliation: 
  - European Molecular Biology Laboratory (EMBL), Genome Biology Unit
- name: Mikhail M. Savitski
  affiliation: 
  - European Molecular Biology Laboratory (EMBL), Genome Biology Unit
date: "`r format(Sys.time(), '%d %B, %Y')`"
package: Rtpca
output:
  BiocStyle::pdf_document:
    keep_md: yes
vignette: >
    %\VignetteIndexEntry{Vignette Title}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8} 
bibliography: bibliography.bib
csl: cell.csl
header-includes: 
- \usepackage{placeins}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
reRunTpca <- TRUE
set.seed(123)
```

# Introduction
Thermal proteome profiling (TPP) [@Savitski2014; @Mateus2020] is a mass spectrometry-based, proteome-wide implemention of the cellular thermal shift assay [@Molina2013]. It was originally developed to study drug-(off-)target engagement. However, it was realized that profiles of interacting protein pairs appeared more similar than by chance [@Tan2018, @Becher2018] which was coined as 'thermal proximity co-aggregation' (TPCA) [@Tan2018]. The R package `Rtpca` enables analysis of TPP datasets using the TPCA concept for studying protein-protein interactions and protein complexes and also allows to test for differential protein-protein interactions (PPIs) across different conditions.      
Here, we exemplify the analysis based on a dataset by @Becher2018 which provides temperature range TPP (TPP-TR) experiments for synchronized HeLa cells in G1/S cell cycle stage versus M phase.             
__Note:__ The paper by @Becher2018 also includes 2D-TPP [@Becher2016] data which is in general more sensitive to changes in protein abundance or stability. This data can also be informative on dynamics of protein-protein interactions based on correlations analysis of 2D-TPP profiles of annotated interactors. However, the advantage of TPP-TR data is that one can test for coaggregation which, if significant, is directly indicative of protein-protein interaction or complex assembly.

# Step-by-step walk through the data analysis

First, we need to load the required libraries (these need to be installed as specified in the comments):

```{r, message=FALSE, warning=FALSE}
library(dplyr) # install.packages("dplyr")
library(readxl) # install.packages("readxl")
library(Rtpca) # BiocManager::install("nkurzaw/Rtpca") 
library(ggplot2) # install.packages("ggplot2")
library(eulerr) # install.packages("eulerr")
```

Then, we download the supplementary data from @Becher2018 which contains the TPP data which we'll be using:

```{r}
if(!file.exists("1-s2.0-S0092867418303854-mmc4.xlsx")){
    download.file(
        url = "https://ars.els-cdn.com/content/image/1-s2.0-S0092867418303854-mmc4.xlsx",
        destfile = "1-s2.0-S0092867418303854-mmc4.xlsx",
        mode = "wb")
}

```

## Getting TPP data into a valid import format
The ideal input format for using the `Rtpca` package is a list of `ExpressionSets` as obtained by
importing the data with the `TPP` package [@franken_2015]. We showcase how this can be done in the `Rtpca` package
vignette.         
However, in order to facilitate usage of the `Rtpca` functions if the raw data of a TPP experiment 
are not available, such as for the case we are exemplifying here, another supported input format of
the data is a simple list of matrices. In the case replicates are available these can be simply incorporated as different list elements. Here we work with the median fold changes for each condition, thus our list objects will only contain one element each. The rows of the matrices should contain the different measured proteins and the columns the relative soluble fraction at different temperatures measured by TMT reporter ions. Since the column names will often represent the measured TMT channels, the `Rtpca` package additionally expects an `attribute` named `temperature` (shown below how to define, not needed if the input is an object imported with the `TPP` package).        

We show here, how such a format can be created based on the example using the supplementary table supplied by @Becher2018:

```{r}
supp_tab_becher_s4 <- read_xlsx("1-s2.0-S0092867418303854-mmc4.xlsx",
                                sheet = "TableS4_TPP-TR")

temperature_anno <- 
    as.numeric(
        gsub("T", "", gsub("_.+", "", colnames(
            supp_tab_becher_s4 %>% 
                dplyr::select(matches("mean\\.fc"))))))
temperature_anno
```
We then extract the data for G1/S:

```{r}
g1s_df <- supp_tab_becher_s4 %>% 
    filter(cell.cycle == "G1_S") %>% 
    dplyr::select(
        gene_name,
        replicates = found.in.reps,
        max_qupm = max.qupm,
        min_qupm = min.qupm,
        matches("mean\\.fc")) %>% 
    filter(min_qupm > 3,
           replicates == 3)
```

And we create as matrix, define its row names and supply and attribute vector specifying the temperatures represented by the various TMT channels:

```{r}
g1s_mat <- as.matrix(
    g1s_df %>% dplyr::select(dplyr::matches("mean\\.fc"))
    )
rownames(g1s_mat) <- g1s_df$gene_name
attributes(g1s_mat)$temperature <- temperature_anno
```

This is what the matrix looks like now:
```{r}
head(g1s_mat)
```

and it's attirbutes
```{r}
summary(attributes(g1s_mat))
```


Now, we do the same for the M phase dataset:
```{r}
m_df <- supp_tab_becher_s4 %>% 
    filter(cell.cycle == "M") %>% 
    dplyr::select(
        gene_name,
        replicates = found.in.reps,
        max_qupm = max.qupm,
        min_qupm = min.qupm,
        matches("mean\\.fc")) %>% 
    filter(min_qupm > 3,
           replicates == 3)
```

```{r}
m_mat <- as.matrix(
    m_df %>% dplyr::select(dplyr::matches("mean\\.fc"))
    )
rownames(m_mat) <- m_df$gene_name
attributes(m_mat)$temperature <- temperature_anno
```

## Multiple testing burden in testing for differential PPIs
In principle, we could now go ahead and test all possible PPIs for differential coaggregation in the datasets acquired in the two different cell cycle phases, however in practise this is not feasible. The reason behind this is that the larger the annotation of PPIs is, the higher will be our multiple testing burden and the less likely we are to identify true positive PPI changes.     
Thus, below we suggest two possible strategies that lead to a significant reduction in tests in comparison to e.g. testing all StringDb [@Szklarczyk2019] annotated PPIs above a certain threshold (even though using a very high threshold (990 or even higher) might also be a viable strategy).      
The first approach ('complex-centric approach') first tests for coaggregation of protein complexes separately in the different conditions. All PPIs in significantly coaggregating complexes in any of the conditions are then used in a secound step to test for differential coaggregation across the conditions.      
The secound approach ('PPI-centric approach') uses a similar strategy, but tests for significant PPI coaggregation separately in the different conditions and then chooses significant interactions across both conditions for further testing for differential behavior.

## Complex-centric analysis

We start by loading an annotation of mammalian complexes by @Ori2016, which comes with the `Rtpca` package:
```{r}
data("ori_et_al_complexes_df")
ori_et_al_complexes_df
```

The crucial columns it contains are the following `protein`: a column using the same identifiers (Gene names (in this case), Uniprot ids or Ensembl protein ids) as the row names of the supplied input matrix or ExpressionSet object and `id`: unique protein complex ids.            

Then, we perform a TPCA analysis based on complexes only in the G1/S condition:

```{r}
G1S_TPCA <- runTPCA(
    objList = list(g1s_mat),
    complexAnno = ori_et_al_complexes_df)
```
We can plot a ROC curve to see how predictive our data is for recovering annotated protein complexes by evoking:

```{r}
plotComplexRoc(G1S_TPCA, computeAUC = TRUE)
```

Since it would not be feasible to compare all annotated protein complexes (true positives) with all non-complex annotated groups of proteins (false positives), the function uses several random permutations of the input complex annotation table as a proxy of false positives found by ranking groups of proteins by low average Euclidean melting curve distances. __Note:__ This procedure will be slightly different each time it is run, since the permutations of the complex annotation table will be different each time. Thus, it is recommended to set a random number generator seed to get reproducible results.           

We can now inspect significantly co-melting protein complexes, like this:
```{r}
tpcaResultTable(G1S_TPCA) %>% filter(p_adj < 0.1) 
```


```{r}
g1s_significant_complex_comelting <- 
    filter(tpcaResultTable(G1S_TPCA), p_adj < 0.1)$complex_name
```

Next, we perform the same analysis for only the M phase condition:

```{r}
M_TPCA <- runTPCA(
    objList = list(m_mat),
    complexAnno = ori_et_al_complexes_df)
```
We can see that the predictive performance of this dataset for protein complexes is not quite as good as for the G1/S one, but still pretty decent: 
```{r}
plotComplexRoc(M_TPCA, computeAUC = TRUE)
```

```{r}
tpcaResultTable(M_TPCA) %>% filter(p_adj < 0.1) 
```

Based on the protein complexes which we find significantly assembled in either condition, we will select the protein-protein interactions to test for in a differential TPCA:
```{r}
m_significant_complex_comelting <- 
    filter(tpcaResultTable(M_TPCA), p_adj < 0.1)$complex_name

all_significant_complex_comelting <- 
    unique(c(g1s_significant_complex_comelting, 
             m_significant_complex_comelting))
```

We load the annotation of protein-protein interactions within complexes that is composed of PPIs from StringDb [@Szklarczyk2019] and the complex annotation by @Ori2016 and filter it for protein complexes that we have seen to coaggregate in the analysis above.

```{r}
data("ori_et_al_complex_ppis")

filtered_complex_ppis <- ori_et_al_complex_ppis %>% 
    filter(complex_name %in% all_significant_complex_comelting) 
```

We now run the differential TPCA by evoking:
```{r eval=reRunTpca}
M_vs_G1S_diff_TPCA <- runDiffTPCA(
    objList = list(g1s_mat),
    contrastList = list(m_mat),
    ctrlCondName = "G1/S",
    contrastCondName = "M",
    ppiAnno = filtered_complex_ppis, 
    n = 10^6
)
```
```{r eval=!reRunTpca, echo=FALSE}
M_vs_G1S_diff_TPCA <- readRDS("prerun/M_vs_G1S_diff_TPCA.rds")
```


This analysis gives us a `tpcaResult` object:
```{r}
M_vs_G1S_diff_TPCA
```


We can now plot the result in form of a volcano plot:

```{r}
plotDiffTpcaVolcano(M_vs_G1S_diff_TPCA, 
                    setXLim = TRUE,
                    xlimit = c(-0.75, 0.75))
```

We can now inspect the significant results:
```{r}
diffTpcaResultTable(M_vs_G1S_diff_TPCA) %>% 
    dplyr::select(pair, rssC1_rssC2, p_value, p_adj) %>% 
    arrange(p_value)
```

To validate significant PPIs we can inspect their melting curves:
```{r}
plotPPiProfiles(M_vs_G1S_diff_TPCA, pair = c("NUP205", "RANBP2"))
```

```{r}
plotPPiProfiles(M_vs_G1S_diff_TPCA, pair = c("RPS6", "RPSA"))
```

## PPI-centric analysis

For the PPI-centric analysis, we first load PPIs annotated by StringDb [@Szklarczyk2019]:
```{r}
data("string_ppi_df")
string_ppi_df
```
This table has been created from the human *protein.links* table downloaded from the StringDb website. It can serve as a template for users to create equivalent tables for other organisms.       
Its essential columns are `x`: 1st interactor, `y`: 2nd interactor and `pair`: unique id of both interactors in alphabetical order.

And we filter this table to only contain high confidence PPIs
```{r}
string_ppi_975_df <- string_ppi_df %>% 
    filter(combined_score >= 975)
```


Then we start our analysis based on PPIs:

```{r, eval=reRunTpca}
G1S_PPI_TPCA <- runTPCA(
    objList = list(g1s_mat),
    ppiAnno = string_ppi_975_df,
    nSamp = 10^6)
```
```{r,  eval=!reRunTpca, echo=FALSE}
G1S_PPI_TPCA <- readRDS("prerun/G1S_PPI_TPCA.rds")
```


As for the complex-centric analysis we get back a `tpcaResult` object:
```{r}
G1S_PPI_TPCA
```

And we can also inspect a ROC curve for this analysis:
```{r}
plotPPiRoc(G1S_PPI_TPCA, computeAUC = TRUE)
```

To inspect which PPIs coaggregated significantly, we can evoke:

```{r}
tpcaResultTable(G1S_PPI_TPCA) %>% filter(p_adj < 0.1) %>% arrange(p_value)
```

And we can run the same analysis for the M-phase dataset:

```{r, eval=reRunTpca}
M_PPI_TPCA <- runTPCA(
    objList = list(m_mat),
    ppiAnno = string_ppi_975_df,
    nSamp = 10^6)
```
```{r, eval=!reRunTpca, echo=FALSE}
M_PPI_TPCA <- readRDS("prerun/M_PPI_TPCA.rds")
```

```{r cache=TRUE}
plotPPiRoc(M_PPI_TPCA, computeAUC = TRUE)
```

__Note:__ The AUC for recovering PPIs using the TPCA approach is usually lower than for recovering protein complexes. This is due to the fact that it is less likely to find three or more proteins showing similar melting curves by chance (protein complex analysis), than it is for two proteins (PPI-based analysis).

```{r}
tpcaResultTable(M_PPI_TPCA) %>% filter(p_adj < 0.1) %>% arrange(p_value)
```


By now combining the significantly found coaggregating PPIs (we are a bit less stringent on the adjusted p-value filter here to not reduce the space of possible differential PPIs too strongly), we can define a list of PPIs which we can use to test for differential PPIs across the two cell cycle phases:
```{r}
ppis_to_test_diff <- unique(
    c(filter(G1S_PPI_TPCA@tpcaResultTable, p_adj < 0.2)$complex_name,
      filter(M_PPI_TPCA@tpcaResultTable, p_adj < 0.2)$complex_name)
)

filtered_string_ppis <- string_ppi_975_df %>% 
    filter(pair %in% ppis_to_test_diff)
```

Based on these PPIs we can now again run a differential TPCA:
```{r, eval=reRunTpca}
M_vs_G1S_PPI_diff_TPCA <- runDiffTPCA(
    objList = list(g1s_mat),
    contrastList = list(m_mat),
    ctrlCondName = "G1/S",
    contrastCondName = "M",
    ppiAnno = filtered_string_ppis, 
    n = 10^6
)
```
```{r, eval=!reRunTpca}
M_vs_G1S_PPI_diff_TPCA <- readRDS("prerun/M_vs_G1S_PPI_diff_TPCA.rds")
```



```{r}
plotDiffTpcaVolcano(M_vs_G1S_PPI_diff_TPCA, 
                    setXLim = TRUE,
                    xlimit = c(-0.75, 0.75))
```

Again, we can now inspect the significant results:
```{r}
diffTpcaResultTable(M_vs_G1S_PPI_diff_TPCA) %>% 
    dplyr::select(pair, rssC1_rssC2, p_value, p_adj) %>% 
    arrange(p_value)
```

And again we plot some of the significantly differentially coaggregating protein pairs:

```{r}
plotPPiProfiles(M_vs_G1S_PPI_diff_TPCA, c("CDC5L", "EXOC7"))
```

```{r}
plotPPiProfiles(M_vs_G1S_PPI_diff_TPCA, c("EIF3D", "EIF3E"))
```

```{r}
plotPPiProfiles(M_vs_G1S_PPI_diff_TPCA, c("CCNT1", "CDK9"))
```
# Comparison of the results obtained by both strategies

In order to asses how many differential PPIs can be recovered with either of the approaches, we plot a Venn diagram below: 

```{r}
complex_centric_hits <- diffTpcaResultTable(M_vs_G1S_diff_TPCA) %>% 
    dplyr::select(pair, rssC1_rssC2, p_value, p_adj) %>% 
    arrange(p_value) %>% 
    filter(p_adj < 0.1)

ppi_centric_hits <- diffTpcaResultTable(M_vs_G1S_PPI_diff_TPCA) %>% 
    dplyr::select(pair, rssC1_rssC2, p_value, p_adj) %>% 
    arrange(p_value) %>% 
    filter(p_adj < 0.1)

all_hits <- union(complex_centric_hits$pair, 
                  ppi_centric_hits$pair)

venn_df <- data.frame(
  complex_centric = all_hits %in% complex_centric_hits$pair,
  ppi_centric = all_hits %in% ppi_centric_hits$pair
)

plot(venn(venn_df))
```
It appears that both approaches pick up a set of distinct differential PPIs. The PPI-centric approach appears to recover more significant PPI changes, however the complex-centric one reveals more intra-complex centered interactions changes.

# Conclusion
`Rtpca` offers user-friendly exploration of TPP datasets for PPIs and allows to assess significantly changing PPIs across different conditions. We exemplify here, how this can be done using the TPP dataset of different phases of the human cell cycle [@Becher2018] from which we recover several differentially coaggregating protein pairs which are known to change during these phases.      
A challenge in the analysis remains the high numbers of hypothesis tests that have to be performed which require multiple testing adjustment and are limited in sensitivity. In the future, methods such as *independent hypothesis weighting* (`IHW`) [@Ignatiadis2016], exploiting covariates such as number of unique peptides used for quantification, could be used to circumvent this problem and further improve sensitivity of the approach.

```{r}
sessionInfo()
```

# References

